staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+	// project). Because capacities are sometimes compared across devices,
+	// a given name is expected to mean the same thing and have the same
+	// type on all devices.
+	//
+	// Capacity names must be either a DNS label
+	// (e.g. "theName") or a DNS subdomain followed by a slash ("/")
+	// followed by a DNS label
+	// (e.g. "example.com/theName"). Capacities whose name do not
+	// include the domain prefix are assumed to be part of the driver's
+	// domain. Capacities defined by 3rd parties must include the domain
+	// prefix.
+	//
+	// The maximum length for the DNS subdomain is 63 characters (same as
+	// for driver names) and the maximum length of the DNS label identifier
+	// is 32.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Quantity determines the size of the capacity.
+	//
+	// +required
+	Quantity *resource.Quantity `json:"quantity,omitempty" protobuf:"bytes,2,opt,name=quantity"`
+}
+
+// DeviceMaxIDLength is the maximum length of the identifier in a device attribute or capacity name (`<domain>/<ID>`).
+const DeviceMaxIDLength = 32
+
+// DeviceAttributeMaxValueLength is the maximum length of a string or version attribute value.
+const DeviceAttributeMaxValueLength = 64
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.31
+
+// ResourceSliceList is a collection of slices.
+type ResourceSliceList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// +optional
+	metav1.ListMeta `json:"listMeta" protobuf:"bytes,1,opt,name=listMeta"`
+
+	// Items is the list of resource slices.
+	Items []ResourceSlice `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.26
+
+// ResourceClaim describes which resources (typically one or more devices)
+// are needed by a claim consumer.
+// Its status tracks whether the claim has been allocated and what the
+// resulting attributes are.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceClaim struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec defines what to allocated and how to configure it.
+	// The spec is immutable.
+	Spec ResourceClaimSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+
+	// Status describes whether the claim is ready for use.
+	// +optional
+	Status ResourceClaimStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ResourceClaimSpec defines how a resource is to be allocated.
+type ResourceClaimSpec struct {
+	// Requests are individual requests for separate resources for the claim.
+	// An empty list is valid and means that the claim can always be allocated
```

Comment:
We went back and forth on this a bit in the KEP discussion, and then I ended up mixing up the two in the API :sad:

My original intent was to allow "empty claims", as described here, because I think it can avoid special cases in "YAML generators": when generating the list, producing an empty claim would be valid. No changes needed in the pod generator, it can always reference the claim.

If we make it required and don't allow empty, then such a generator has avoid creating and referencing empty claims.

I don't have a strong opinion on this. If I don't hear otherwise, then I'll make it required and remove this comment here.

Reply:
Done.